Improving Metro Support

kimchouard opened the pull request:
- BUGFIX on `Platform.web.tsx` to account for potential `undefined` value for `window`
- Updated documentation for code-splitting, that won't work when using metro (production build will fail if using `import(...)` but works great with `require(...)`

It's small changes but that can have a big impact and got my builds to fail until I found the fix. Open for feedback if I missed anything. :)

wcandillon commented on docs/docs/getting-started/web.mdx:
````diff
@@ -174,7 +181,7 @@ export default function App() {
+      getComponent={() => import("./MySkiaComponent")} // use require(...) if you're using the metro bundler
   );
 }
 ```
````
Looks like there is something wrong here?

kimchouard replied:
It _looks like_ it indeed, but it's having the same odd behavior in the [current `web.mdx` page](https://github.com/Shopify/react-native-skia/blob/main/docs/docs/getting-started/web.mdx):
<img width="1541" alt="CleanShot 2023-11-30 at 16 48 44@2x" src="https://github.com/Shopify/react-native-skia/assets/2856923/a09bc0ec-4566-4593-9e07-436ecc16d90b">

I've really just added a comment `//...` and nothing more (like line 136), so that should work fine.

kimchouard commented:
Alright, I've tried to make it as simple as possible for the end-user. Running `yarn setup-skia-web` should auto-detect if metro is used with Expo and act accordingly. ✨

Changelog:
- I've updated the `setup-canvaskit.js` script to take Expo Metro into account (if detected). The script output should look something like this now:
<img width="805" alt="CleanShot 2024-01-05 at 19 53 56@2x" src="https://github.com/Shopify/react-native-skia/assets/2856923/5ae00146-6c3e-4b96-9622-48ce0cf2be4c">

- I've also updated the docs accordingly.

To test:
```
# Create a boilerplate if you don't have a codebase already
npx create-expo-app@latest --template tabs@49

# Install React Native Skia
expo install @shopify/react-native-skia
yarn setup-skia-web # also add it to the postinstall script in package.json
```

@wcandillon let me know your thoughts. It should fix a few of the issues opened right now. 🙏

omeid commented:
@kimchouard I don't see that error, and here is how I load skia:

```tsx
import { WithSkiaWeb } from '@shopify/react-native-skia/lib/module/web'
import { Text } from 'react-native'

const Loader = () => (
  <WithSkiaWeb
    opts={{
      locateFile: () => '/assets/canvaskit.wasm',
    }}
    getComponent={() => import('../editorWithContext')}
    fallback={<Text>Loading...</Text>}
  />
)

export default Loader
```

omeid commented:
We do have a custom build of canvaskit to include some bug fixes that aren't released tho.

kimchouard commented:
> We do have a custom build of canvaskit to include some bug fixes that aren't released tho.

Ah, that might be why then 🤔

Just checking: you're compiling with Metro not webpack right?

kimchouard commented:
> @kimchouard Yes, metro.
> 
> ```json
>     "web": {
>       "bundler": "metro",
>       "favicon": "./assets/favicon.png"
>     },
> ```
> 
> And I also tried a the latest canvaskit.wasm from npm, it works just as well.

Ha, bizarre... 🤨 Last check, you're testing the **compiled** version (for prod) right ?

It works great for me in dev mode (`npx expo start`) but crashed after minification with `npx expo export --platform web`.

omeid commented:
> Ha, bizarre... 🤨 Last check, you're testing the compiled version (for prod) right ?

Yes, build, and then test with `python -m http.serve` (not that the http server should matter much).

node v20.9.0
pnpm 8.14.1
linux
@shopify/react-native-skia 0.1.225

.npmrc with `node-linker=hoisted` as only config.

kimchouard commented:
Thanks for sharing @omeid. Metro, unlike Webpack, does not support code splitting out-of-the-box, so I'm curious what in your config makes it work. Can you share your `metro.config.js`, `babel.config.js` and `tsconfig.json` please?

When using official templates, the issues is **_predictably_ reproduceable** (I've confirmed that on different machines from different devs, e.g. @oliviergraziano), simply by following the getting started steps with Metro/Expo Router (_see below_). Using `require` instead `import` consistently fixes the issue, and the updated `setup-skia-web` script makes the setup a breeze for **both** Webpack and Metro. 🌬️❤️

@wcandillon open to your feedback here, if I'm missing something or if anything should be updated for this PR to be merged. 🙂

---

## Steps to reproduce the issue (fixed in this PR, both via script & documentation)

### 1. Start from the boilerplate

```bash
# Create a new app from Expo Router boilerplate
npx create-expo-app@latest --template tabs@49

# Then go to your app directory and install all the dependencies
cd my-app 
yarn install

# Install React Native Skia
yarn add @shopify/react-native-skia
yarn setup-skia-web # (1) see note below 
```

**(1) Note:** `yarn setup-skia-web` should work out of the box for Metro/Expo with my PR. 
👉 If you're using current RN-skia version, you'll need to run `yarn setup-skia-web \"public\" && node postinstall.js` (using the `postinstall.js` script from #1947).

### 2. Create an [`HelloWorld` Skia](https://shopify.github.io/react-native-skia/docs/getting-started/hello-world) component

For example in `components/HelloSkia.tsx`:

```tsx
import React from "react";
import { Canvas, Circle, Group } from "@shopify/react-native-skia";
 
const HelloSkia = () => {
  const width = 256;
  const height = 256;
  const r = width * 0.33;
  return (
    <Canvas style={{ width, height }}>
      <Group blendMode="multiply">
        <Circle cx={r} cy={r} r={r} color="cyan" />
        <Circle cx={width - r} cy={r} r={r} color="magenta" />
        <Circle cx={width / 2} cy={width - r} r={r} color="yellow" />
      </Group>
    </Canvas>
  );
};
 
export default HelloSkia;
```

### 3. Load Skia with Code-Splitting

Update a page to use `<WithSkiaWeb .../>`. For example, `app/(tabs)/two.tsx` becomes:

```tsx
import { Text } from 'react-native';
import { WithSkiaWeb } from '@shopify/react-native-skia/lib/module/web';

export default function TabTwoScreen() {
  return (
    <WithSkiaWeb
      opts={{ locateFile: (file) => `static/js/${file}` }}
      getComponent={() => import('../../components/HelloSkia')} // this will fail on production build
      fallback={(<Text>Loading</Text>)}
    />
  );
}
```

### 4. Run the code

```bash
# Works in development
npx expo start --web

# But it fails in prod 
npx expo export --platform web
npx serve dist # or `python -m http.serve`
```

Production code will throw a `Requiring unknown module: XX` when loading the 2nd tab with the Skia code:
<img width="595" alt="CleanShot 2024-01-23 at 17 42 55@2x" src="https://github.com/Shopify/react-native-skia/assets/2856923/e6c54fbb-db90-4495-a627-7b5815a4d582">

### 🥳 Proposed FIX: replace `import` by `require`

Replace the dynamic `import` to use `require` in `<WithSkiaWeb .../>` and re-build using `npx expo export --platform web`.

If you load the built code from the `dist/` folder (e.g. via `npx serve dist`), it will now work! 🍾

```tsx
<WithSkiaWeb
  opts={{ locateFile: (file) => `static/js/${file}` }}
  getComponent={() => require('../../components/HelloSkia')} // this will work, even on the production build
  fallback={(<Text>Loading</Text>)}
/>
```

kimchouard commented:
> @kimchouard Can you try pnpm please?

Do you mean [this](https://pnpm.io/)? 🤔 Anything specific I should do with it?

kimchouard commented:
> Just use it instead of yarn. :)

Hehe nice, yes I'll give it a shot. Thanks ;)
👉 Though we still need to support `npm` and `yarn` for the greater community (:

kimchouard commented:
Closing this PR in favor of a more global one on the RN Skia's docs: #2176
